packages/test/end-to-end-tests/src/test/codePropsal.spec.ts

```diff
@@ -0,0 +1,148 @@
+    let container2: Container;
+    beforeEach(async () => {
+        deltaConnectionServer = LocalDeltaConnectionServer.create();
+
+        // Create a Container for the first client.
+        container1 = await createContainer([["map", SharedMap.getFactory()]]);
+
+        opProcessingController = new OpProcessingController(deltaConnectionServer);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        await opProcessingController.process();
+
+        // Load the Container that was created by the first client.
+        container2 = await loadContainer([["map", SharedMap.getFactory()]]);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        const quorum1 = container1.getQuorum();
+        const quorum2 = container2.getQuorum();
+
+        assert.deepStrictEqual(
+            quorum1.get("code"),
+            codeDetails,
+            "Code proposal in container1 doesn't match");
+
+        assert.deepStrictEqual(
+            quorum2.get("code"),
+            codeDetails,
+            "Code proposal in container2 doesn't match");
+
+        const dataObject1 = await requestFluidObject<ITestFluidObject>(container1, "default");
+        const map1 = await dataObject1.getSharedObject<ISharedMap>("map");
+
+        // BUG BUG quorum.propose doesn't handle readonly, so make sure connection is write
+        while (container1.deltaManager.connectionMode === "read" || !container1.connected) {
+            map1.set("foo","bar");
+            await Promise.all([
+                new Promise((resolve) => container1.connected ? resolve() : container1.once("connect", resolve)),
+                opProcessingController.process(),
+            ]);
+        }
+    });
+
+    it("Code Proposal", async () => {
+        container1.on("contextChanged",(c)=>{
+            assert.deepStrictEqual(
+                c,
+                codeDetails2,
+                "container1 context should be update");
+        });
+
+        container2.on("contextChanged",(c)=>{
+            assert.deepStrictEqual(
+                c,
+                codeDetails2,
+                "container2 context should be update");
+        });
+
+        await Promise.all([
+            container1.getQuorum().propose("code", codeDetails2),
+            opProcessingController.process(),
+        ]);
+    });
+
+    it("Code Proposal Rejection", async () => {
+        const quorum1 = container1.getQuorum();
+        const quorum2 = container2.getQuorum();
+
+        container1.on("contextChanged",(c)=>{
+            assert.fail("Conext Shouldn't Change for container1");
+        });
+
+        container2.on("contextChanged",(c)=>{
+            assert.fail("Conext Shouldn't Change for container2");
+        });
+
+        quorum2.on("addProposal",(p)=>{
+            if (p.key === "code") {
+                p.reject();
+            }
+        });
+
+        await Promise.all([
+            quorum1.propose("code", codeDetails2)
```

Comment:
I think it should propose something other than codeDetails2, or we should have more resetting happening between tests. Since it is the exact same instance of pkg, I think there's a chance that even with accepting the code proposal, contextChanged event would not be raised, because the code in the Container event handler ignores the accepted code proposal if it is the same as the current one.

Reply:
Oh I see, it is taken care of by the `beforeEach`. Disregard!